platform/pulp/download/writer.py

```diff
@@ -0,0 +1,144 @@
+import os
+import errno
+
+from io import BytesIO
+from logging import getLogger
+
+
+log = getLogger(__name__)
+
+
+class Writer:
+    """
+    Downloaded content writer.
+
+    Attributes:
+        path (str): The absolute path to the file. May be None.
+        fp: An open file-like object used for writing.
+    """
+
+    def __init__(self, path=None, fp=None):
+        """
+        Attributes:
+            path (str): The absolute path to the file. May be None.
+            fp: An open file-like object used for writing.
+        """
+        self.path = path
+        self.fp = fp
+
+    @property
+    def is_open(self):
+        """
+        Get whether the writer is open.
+
+        Returns:
+            bool: True when open.
+        """
+        return self.fp is not None
+
+    def open(self):
+        """
+        Open for writing.
+        The directory tree is created as necessary.
+        """
+        pass
+
+    def append(self, bfr):
+        """
+        Append (write) the buffer.
+
+        Args:
+            bfr (bytes): A buffer to append.
+
+        Returns:
+            int: The number of bytes appended.
+        """
+        return self.fp.write(bfr)
+
+    def close(self):
+        """
+        Close the target file.
+        """
+        pass
+
+    def discard(self):
+        """
+        Discard written content.
+        """
+        pass
+
+    def __enter__(self):
+        self.open()
+        return self
+
+    def __exit__(self, *unused):
+        self.close()
+
+
+class FileWriter(Writer):
+    """
+    Downloaded content writer.
+    Write content to a file.
+    """
+
+    def open(self):
+        """
+        Open for writing.
+        The directory tree is created as necessary.
+        """
+        if self.is_open:
+            return
+        self._mkdir()
+        self.fp = open(self.path, 'wb')
+
+    def close(self):
+        """
+        Close the target file.
+        """
+        if not self.is_open:
+            return
+        try:
+            self.fp.close()
+        except Exception:
+            log.exception(self)
+        finally:
+            self.fp = None
+
+    def discard(self):
+        """
+        Discard written content.
+        """
+        os.unlink(self.path)
+
+    def _mkdir(self):
+        """
+        Create the directory as needed.
+
+        Raises:
+            OSError: When the directory cannot be created.
+        """
+        _dir = os.path.dirname(self.path)
+        if not _dir:
+            return
+        try:
+            os.makedirs(_dir)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise
+
+    def __str__(self):
+        return self.path
+
+
+class TextWriter(Writer):
+    """
+    Downloaded text writer.
+    Used to store downloaded text in memory.
+    """
+
+    def open(self):
+        self.fp = BytesIO()
+
+    def __str__(self):
+        self.fp.seek(0)
+        return self.fp.read().decode('utf8', 'replace')
```

Comment:
Fair point.  The intended usage is to support conversion to a ``str`` by doing:
```
s = str(writer)
```
more than a to _"to compute the “informal” or nicely printable string representation of an object"_ as defined in the python3 docs.  I originally named it `to_str()` but this seems even odder because is raised the question of, why not impl as ``__str__()`` which is the standard conversion to a ``str``. 

The main goal here is to provide user a convenient way to get the data captured by the TextWriter to  ``str`` and use it.  If you still feel strongly, I'm open to renaming the method.